src/ic-cdk/src/api/stable.rs

```diff
@@ -161,6 +162,142 @@ impl io::Write for StableWriter {
+        BufferedStableWriter {
+            offset: 0,
+            capacity: memory.stable64_size(),
+            buffer: Vec::with_capacity(buffer_size),
+            memory,
+        }
+    }
+
+    /// Writes a byte slice to the buffer, flushing the buffer to stable memory if it becomes full.
+    ///
+    /// The only condition where this will error out is if it cannot grow the memory.
+    pub fn write(&mut self, buf: &[u8]) -> Result<usize, StableMemoryError> {
+        let buffer_capacity_remaining = self.buffer.capacity() - self.buffer.len();
+
+        match buffer_capacity_remaining.cmp(&buf.len()) {
+            // There is enough room in the buffer to store the new bytes.
+            Ordering::Greater => {
+                self.buffer.extend_from_slice(buf);
+            }
+            // There is enough room in the buffer to store the new bytes, but now it is full so we
+            // need to flush it to stable memory.
+            Ordering::Equal => {
+                self.buffer.extend_from_slice(buf);
+                self.flush()?;
+            }
+            // The new bytes will not fit in the buffer.
+            Ordering::Less => {
+                // We can reduce the calls to grow stable memory by growing to the total known
+                // length here rather than leaving it up to `flush` which will only grow by up to
+                // the length of the buffer.
+                let total_capacity_required =
+                    self.offset + self.buffer.len() as u64 + buf.len() as u64;
+                self.grow_to_capacity_bytes(total_capacity_required)?;
+
+                // If the new bytes exceed the capacity remaining + the starting capacity then we
+                // must flush everything now since we will not be able to fit the bytes into the
+                // buffer.
+                // Otherwise we fill the buffer, flush it, then populate the buffer again with the
+                // remaining bytes.
+                if buf.len() > self.buffer.capacity() + buffer_capacity_remaining {
+                    self.flush()?;
+                    self.memory.stable64_write(self.offset, buf);
+                    self.offset += buf.len() as u64;
+                } else {
+                    self.buffer
+                        .extend_from_slice(&buf[..buffer_capacity_remaining]);
+                    let remaining_to_write = &buf[buffer_capacity_remaining..];
+                    self.flush()?;
+                    self.buffer.extend_from_slice(remaining_to_write);
+                }
+            }
+        }
+
+        Ok(buf.len())
+    }
+
+    /// Attempts to grow the memory by adding new pages.
+    pub fn grow(&mut self, added_pages: u64) -> Result<(), StableMemoryError> {
+        let old_page_count = self.memory.stable64_grow(added_pages)?;
+        self.capacity = old_page_count + added_pages;
+        Ok(())
+    }
+
+    pub fn flush(&mut self) -> Result<(), StableMemoryError> {
```

Comment:
Maybe mention in the comment that this is a relatively expensive operations as it going to grow the memory.
So that the user doesn't call flush after each write.

Reply:
Done.